### ISSUE TITLE
DBBC3 support in DRUDG

### ISSUE THREAD

**Issue by varenius** [nvi-inc/fs#33]
To what extent is DBBC3 supported in DRUDG? If non/limited; do we know what is lacking?

**Comment by wehimwich** [comment]
It is only supported for VGOS (in **.skd** files)  for now, i.e., 10.0.0 (and 9.12.13).  That only works  because _drudg_ doesn't generate the procedures for **BB** rack systems now. It takes them from explicit listings in the **.skd** file. So any combination of FS recorder and DBBC3 commands, including **dbbc3=...** low-level commands, can be used. 

(If you have **.prc**  files that work for Oe/Ow, we could include them in schedules, as we do for RDBE stations, so that there is a repeatable, schedule/mode specific, set-up and so _drudg_-ing is more "normal".)

For general support with **.vex** schedules, @gipson55 is the expert, but we probably need at least four things in _drudg_:

- [ ]  Being able to select DBBC3 as a rack type
- [ ]  Larger arrays for IFs and BBCs (maybe also channels), and accepting a larger population on read-in
- [ ]  Formatted output (based on DBBC2 output) for the large number of components  in **.prc** output
- [ ]  Output of other necessary setup commands (e.g., VDIF output definition) for DBBC3 and recorder. It probably would not be possible to specify the set-up fully in the schedule until VEX2, but there might be some useful standard cases we could detect.

Until the setup commands are defined, they could be left out if there are other means of getting the same things set (maybe generating explicit **dbbc3=...** commands for now, but if that won't work, an external client might be an option; maybe the latter is what is being done now). Ultimately they need to be defined and included.

For up to 16  BBCs  and 4 IFs (in **.skd** or **.vex**),  you might be able to fudge to use a DBBC3 in place of a DBBC2 in a schedule with _drudg_ i.e.,, 10.0.0 (and 9.13.2), if you change the contents of a DBBC2 **.prc** file to:

- Adjust format of the **bbcXX=...** commands
- Adjust format of the **ifX=...** commands
- Remove extraneous DBBC2 commands
- Possibly add some other commands you might want (e.g, **dbbc3=...**) for other DBBC3 setup (VDIF output etc).

Also you might need an external client if you can't do everything with regular and **dbbc3=...** commands.

Supporting just 16 BBCs  and 4 IFs in generally might work if the above changes for a DBBC2 output are triggered by selecting a DBBC3 rack type. It might not be possible to detect all of the information for needed  **dbbc3=...** commands,  but we could see. There might be some standard setups that could be recognized. Ultimately, this is probably the best that could done for **.skd** files, again @gipson55 is the expert.

**Comment by varenius** [comment]
I'm testing DRUDG with the ed/dbbc3 branch, trying to get it to run for DBBC3. Some relevant file conents:
**/usr2/control/equip.ctl has**
```
dbbc3_ddc_v
*mk3     type of rack (mk3, vlba, vlbag, mk4, vlba4, mk5, vlba5
*                     k41, k41u, k41/k3, k41u/k3, k41/mk4, k41u/mk4,
*                     k42, k42a, k42b, k42bu, k42c, k42/k3, k42a/k3,
*                     k42bu/k3, k42/mk4, k42a/mk4, k42b/mk4, k42bu/mk4
*                     k42c/mk4, lba, lba4, s2, dbbc_ddc, dbbc_ddc/fila10g,
*                     dbbc_pfb, dbbc_pfb/fila10g, vlbac, cdas, rdbe, dbbc3,
*                     or none)
flexbuff   type of recorder 1 (mk3, vlba, vlba2, vlbab, vlba4, vlba42, mk4,
*                   mk4b, s2, k41, k41/dms, k42, k42/dms, mk5a, mk5a_bs,
*                   mk5b, mk5b_bs, mk5c, mk5c_bs, flexbuff, mk6, or none)
none    type of recorder 2 (mk3, vlba, vlba2, vlbab, vlba4, vba42, mk4,
*                           mk4b, or none)
none     type of decoder (mk3, dqa, mk4, or none)
```
/usr2/control/skedf.ctl has:
```
* Station equipment:
* Station equipment may be specified in drudg or in the equip line below.
*  Equipment names are NOT case sensitive.
*  Allowed rack and recorder names are:
*  Racks        |  Recorders
*  -------------+-----------
*  none         |  none
*  unknown      |  unknown
*  Mark3A       |  unused
*  VLBA         |  Mark3A
*  VLBAG        | VLBA
*  VLBA/8       | VLBA4
*  VLBA4/8      | Mark4
*  Mark4        | S2
*  VLBA4        | K4-1
*  K4-1         | K4-2
*  K4-2         | Mark5A
*  K4-1/K3      | Mk5APigW
*  K4-2/K3      | Mark5P
*  K4-1/M4      | K5
*  K4-2/M4      | Mark5B
*  LBA          | Mark5C
*  Mark5        | Flexbuff
*  VLBA5        |
*  DBBC_ddc     |
*  DBBC_ddc/FILA10G |
*  DBBC_pfb     |
*  DBBC_pfb/FILA10G |
*  VLBAC        |
*  CDAS         |
*  BB           | Mark6
* DBBC3_DDC
DBBC3_DDC
```
If I run "drudg vo1259 Ow 12 0" I get a PRC file assuming a DBBC2 backend, i.e. it says:
```
 NOTE: These procedures are for the following equipment:
   Rack:       DBBC_DDC
   Recorder 1: FlexBuff
```
 This is surprising, because I would assume it should read the DBBC3 rack type from either equip.ctl or skedf.ctl?

If I run "drudg vo1259" and select antenna "Ow" and then "11" to select rack type, I get the option list
```
ONSA13SW equipment: Rack=DBBC_DDC            Recorder=FlexBuff
| Select rack          | Select Rec 1 |
|   1=none             |   1=none     |
|   2=Mark3A           |   2=unused   |
|   3=VLBA             |   3=Mark3A   |
|   4=VLBAG            |   4=VLBA     |
|   5=VLBA/8           |   5=VLBA4    |
|   6=VLBA4/8          |   6=Mark4    |
|   7=Mark4            |   7=S2       |
|   8=VLBA4            |   8=K4-1     |
|   9=K4-1             |   9=K4-2     |
|  10=K4-2             |  10=Mark5A   |
|  11=K4-1/K3          |  11=Mk5APigW |
|  12=K4-2/K3          |  12=Mark5P   |
|  13=K4-1/M4          |  13=K5       |
|  14=K4-2/M4          |  14=Mark5B   |
|  15=LBA              |  15=Mark5C   |
|  16=Mark5            | *16=FlexBuff |
|  17=VLBA5            |  17=Mark6    |
|  18=RDBE             |              |
| *19=DBBC_DDC         |              |
|  20=DBBC_DDC/Fila10g |              |
|  21=DBBC_DDC/VSI2    |              |
|  22=DBBC_PFB         |              |
|  23=DBBC_PFB/Fila10g |              |
|  24=DBBC3_DDC        |              |
|  25=VLBAC            |              |
|  26=CDAS             |              |
|  27=BB               |              |
```
If I here select "24" for DBBC3, and then generate PRC file, then it seems that **it actually works**! I get reasonable core3h commands etc. (Assuming S/X setup, because there's no proper VGOS in the SKD file for Ow yet). 

**Final notes:**
- Great that DRUDG seems to be able to generate PRCs for DBBC3!
- Why isn't my equip.ctl/skedf.ctl preset for DBBC3 being recognized by DRUDG?

**Comment by jfhquick** [comment]
You also have to uncomment the entry

equipment_override

in skedf.ctl before DRUDG will use the equipment specified there in preference
to what is defined for your station in the VEX file.

**Comment by varenius** [comment]
Apologies, I forgot to mention that: I have already uncommented this line, so it looks like
```
* If equipment_override is specified (uncommented below) then the
* equipment in the control file is used. This then becomes your default
* equipment regardless of what is in the schedule. This is a useful way of
* forcing the recorder to be Mark5A during the transition from tape to disk.
equipment_override
```
So it seemd DRUDG *should* override it, but doesn't.

**Comment by wehimwich** [comment]
I think the issue may be that in _skedf.ctl_ you have:

```
DBBC3_DDC
```

instead of (case should not matter):

```
equipment dbbc3_ddc flexbuff none
```

The leading comment  in what @varenius quoted from _skedf.ctl_ mentions `equip line below`, it should say `equipment line below`, but all the provided examples use `equipment`. There is a tendency to spell words out completely for program input. I have fixed the comment and push it onto the branch. No one needs to `pull` just for that since it doesn't affect functionality.

_drudg_ only gets the equipment from (in priority order):

1. option `11`
2. _skedf.ctl_, with `equipment_override`
3. _skedf.ctl_, without `equipment_override`, if there is no equipment in the schedule
4. the schedule

There is no use of _equip.ctl_.

I have checked off all the items in the task list above. They are either complete or as complete as they can be at this time. Any issues remaining are more like bugs to be fixed.

**Comment by varenius** [comment]
Thanks @whimwich , my mistake with the equipment line was indeed the reason here. (Sorry, I merged old and new skedf.ctl and mistyped this line). Now I get
Warning! All 2nd recorders set to 'none'
 WARNING! Using equipment from skedf.ctl:
     Replacing rack DBBC_DDC             by DBBC3_DDC           
     Replacing rec1     FlexBuff by FlexBuff    
Procedures for ONSA13SW
 PROCEDURE LIBRARY FILE /usr2/proc/vo1259ow.prc FOR ONSA13SW
 NOTE: These procedures are for the following equipment:
   Rack:       DBBC3_DD
   Recorder 1: FlexBuff
   Defining:   exper_initi setupsx dbbcsx8 ifdsx core3hsxDRUDG DONE

- I note that the "Rack" seems to truncate away the last C from DBBC3_DDC. Bug or feature? :)
- I also wonder about https://github.com/nvi-inc/fs/blob/ed/dbbc3/st.default/control/skedf.ctl#L323. It may be wise to have "32000" for all the 8 IFs as template? The reason is this message I got from Sven last year when I asked about the target levels:
> On 2021-05-06, 13:45, "dornbusch" wrote:
> the IF target level should always be 32k for a 0-4 GHz input, sometimes it might be required to reduce it if you have an inputband of lower bandwidth. If you go higher the samplers will go into saturation. This is true for all observation modes.

**Comment by wehimwich** [comment]
Thanks.

> I note that the "Rack" seems to truncate away the last C from DBBC3_DDC. Bug or feature? :)

Probably both.

We can start a list here of things to fix in _drudg_:

- [ ] `Rack:       DBBC3_DD` when making  _.prc_ is missing the trailing `C`
- [ ] Seems like we could use a space before `DRUDG` in `Defining:   exper_initi setupsx dbbcsx8 ifdsx core3hsxDRUDG DONE`

No stoppers yet.

I fixed the `32000` values in the example and pushed it and the fix for #142 in one commit. The other values were left over from testing.  Still nothing that affects functionality, so no urgency to pull.

**Comment by wehimwich** [comment]
For the _10.1_ release: The first checkbox was completed. The third checkbox was completed, but its content was changed. Now `fb` is used in place of `mk5` for all commands and procedures when a FlexBuff is in use. The second checkbox was missed in the latest round of changes. It will be addressed if we need to open _drudg_ up again before the official _10.1_ release. Otherwise, I am going with that being too minor a cosmetic issue to warrant redoing the laborious update and final testing. The next commit for _drudg_ will close this issue.  If you prefer, feel free to open a new issue for the missed checkbox.
